README.md
# CDN Template in NodeJS

[![CircleCI](https://circleci.com/gh/dan-online/cdn-template.svg?style=svg)](https://circleci.com/gh/dan-online/cdn-template)

## Setup

`npm run setup`

`npm install`

`npm start`

## Authors

- [Dan](https://github.com/MayorChano)

## Features

- Image sizing and quality

- Caching

- Minifying results

## How to use

1. Upload files to [src/public/images](src/public/images)
2. Start with `npm start`
3. Go to localhost:8080/f/file_name.png
4. Use **q** for quality, **w** for width and **h** for height Ex.
   `localhost:8080/f/file_name.png?h=200&w=200&q=100`

## License

(The MIT License)

Copyright (c) 2019 DanCodes <dan@dancodes.online>

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


docs/CHANGELOG.md
# nodejs v5.1.0
 - [8715748](../../commit/8715748fe4454e7778272ea1aabab0b8f58c9174)     -      changelog
 - [435bebe](../../commit/435bebecdbe2cb3927fe4cb059f4a9db65cc25a5)     -      changelog
 - [a2fb918](../../commit/a2fb918d7c905278c225c2db9f77e13b70c8b7a8)     -      Merge branch 'master' of https://github.com/MayorChano/NodeJS
 - [fd6fe76](../../commit/fd6fe76bb042b82f0c7f0545831734b42e3fb202)     -      5.1.0
 - [a656f61](../../commit/a656f61b07389eae920687e0c85763f52761e8fa)     -      5.0.0
 - [7f3d3d5](../../commit/7f3d3d5b21fcb7244da456ec66d526cd72667302)     -      Delete CHANGELOG.md
 - [1996971](../../commit/19969712ac7e5f822c3ee3da2754ba29049d854d)     -      Update .gitignore
 - [d993b5c](../../commit/d993b5cfb906db456413c049b06d3479b74d2b0b)     -      Delete package-lock.json
 - [24a8c4e](../../commit/24a8c4e53a1d1e61098a9ef57a2b70ae4f202ec2)     -      Create FUNDING.yml
 - [7c679f8](../../commit/7c679f84ca21c7483f7a749cadc51ffc845541bc)     -      database is already online
 - [91de4fd](../../commit/91de4fdcf9feeb028db24b6ceca316692e66dae8)     -      keywords
 - [2330e21](../../commit/2330e21b9223d60a48c94a464685500fddc7125f)     -      oops
 - [4753582](../../commit/47535826b4ca54b15b3be4548a023c6b78df1069)     -      one parentheses fix
 - [3ed2120](../../commit/3ed212001f2e5a94ed39dedd9ec7002db8097348)     -      so many codacy
 - [6b07448](../../commit/6b07448a9db4ef984887d9cc48f5f88906fdfbd8)     -      double-quotes
 - [347f32d](../../commit/347f32d0cc7edb1e95909f6e88c88a5c5a2a7fa2)     -      fix double quoes
 - [9dd5135](../../commit/9dd5135ca415c7572cb928c943e85e3c777cf285)     -      stuff
 - [729a93f](../../commit/729a93f36ff49bf1ae6c6bc440e4023953ad732b)     -      fixes
 - [8384cb2](../../commit/8384cb23d1a03d2586e77414610079dabd0f1036)     -      amazing looking startup
 - [da4d102](../../commit/da4d102f29797d3ca0e5df6df1737834bb2e8ad8)     -      Merge pull request #9 from MayorChano/dependabot/npm_and_yarn/express-4.17.1
 - [4e649a8](../../commit/4e649a826fdbd11bb23337a3d7e5bc55da5b95e0)     -      Bump express from 4.17.0 to 4.17.1
 - [c139075](../../commit/c139075b39e06feb323b16837dc22eb8a5b104ea)     -      Merge pull request #6 from DivySrivastava/master
 - [6d15831](../../commit/6d15831aeb89a125600fa1d8cbd6e9a77f19fdd0)     -      Update document.js
 - [ca293ba](../../commit/ca293ba232c5f05b9dde4c71a124481009813e00)     -      Update database.js
 - [eb57533](../../commit/eb57533b072a1374a74be13286cd06e034d9b5fd)     -      Merge branch 'master' into master
 - [2ed93c1](../../commit/2ed93c1c6e5c638aa89d3abc7ad9efb039be23ce)     -      Merge pull request #8 from MayorChano/dependabot/npm_and_yarn/mongoose-5.5.11
 - [958fa4a](../../commit/958fa4ab203dc8832c5889f13263c29f03090263)     -      Bump mongoose from 5.5.10 to 5.5.11
 - [663f8d3](../../commit/663f8d3b6305606c56bac8b29ac2dabc94842a57)     -      Merge pull request #7 from MayorChano/dependabot/npm_and_yarn/mongodb-3.2.6
 - [23d99fd](../../commit/23d99fd0dd95a830278d5eff9a2fc9542e637330)     -      Bump mongodb from 3.2.5 to 3.2.6
 - [2aaae19](../../commit/2aaae191716878a89e3e4f193500b625594ea3c4)     -      Add mongoose model
 - [174a090](../../commit/174a090af377f2e5ae7bcb2907a795ac4d6b1b28)     -      Merge pull request #5 from MayorChano/dependabot/npm_and_yarn/express-4.17.0
 - [a52c7ec](../../commit/a52c7ec8679550a1280163226cd2ad3ecfb58d10)     -      Merge pull request #4 from MayorChano/dependabot/npm_and_yarn/mongodb-3.2.5
 - [d5f488f](../../commit/d5f488f29d1eb5728501943fdb9b331239db866e)     -      Bump mongodb from 3.2.4 to 3.2.5
 - [ad8844a](../../commit/ad8844ab79936a32369ae4c35eb171fe2914ed17)     -      Merge pull request #3 from MayorChano/dependabot/npm_and_yarn/mongoose-5.5.10
 - [c75c536](../../commit/c75c536631c8397027ac30ed77b6647555d77274)     -      Bump express from 4.16.4 to 4.17.0
 - [4b3427e](../../commit/4b3427ea9d87bcfa96f7a40853779d4a4c253da5)     -      Bump mongoose from 5.5.8 to 5.5.10
 - [edc2264](../../commit/edc22642ffcb5cb83fc90118c75ede417d0ff73b)     -      double quotes
 - [2e8baf8](../../commit/2e8baf8991db3e99cbb1bb917bd103ef0913e93c)     -      semi-colons
 - [419e8fa](../../commit/419e8fa828cf749efbeb31b0e3d29b806df787ff)     -      eco-system
 - [74cfc90](../../commit/74cfc905213ca9855a5464677ce17f236dbc7c47)     -      uneeded packages
 - [58054a4](../../commit/58054a402751b77dd9237872cf5f60cc530be1d4)     -      4.0.0
 - [809e809](../../commit/809e809d5530631af0dd415f624ce7871e4d90a2)     -      say its done
 - [c430256](../../commit/c4302566f61fd95fe2c390abd7cd060f9d28d653)     -      fix stuff
 - [c7981cd](../../commit/c7981cd06a2704182246ce77805892aeb5060b7c)     -      fixes
 - [89897c0](../../commit/89897c0df5043fe4d7adaa86bdbaca26cadc79ec)     -      data
 - [582c75d](../../commit/582c75d4c83b24cbb60f198893f3b94db58cf424)     -      3.1.10
 - [c77e7ba](../../commit/c77e7ba8973d97497ff123e5ad000756617ca003)     -      Fix
 - [0d31705](../../commit/0d31705292018a24b358da7788f0665cd0e16fdb)     -      basge
 - [fd2a33f](../../commit/fd2a33fa5fafcdf872d65b86b7e846572f644a19)     -      double quotes
 - [a7d9597](../../commit/a7d9597a434773a7ecbf4a5919d03272ec907a3a)     -      setup syntax
 - [df731f0](../../commit/df731f0618d5b2147366a11cf582605561e39e49)     -      Codacy first half
 - [5a9db34](../../commit/5a9db34e6c27ec9efca0024c49339d657e63783d)     -      Changelog
 - [7a31aea](../../commit/7a31aea2e7cc61d5cbb5525b246f7732bb3bb16c)     -      uninstall sentry
 - [324d148](../../commit/324d14885f65da19a54abaf573936cb851a064b8)     -      I won't
 - [d701a08](../../commit/d701a08746b75f1427d68a95e309992e586313e6)     -      sentry
 - [6041c3c](../../commit/6041c3ce37ce8ad9a5df39c34eb2cb6eccb2bf51)     -      setup
 - [ed822f0](../../commit/ed822f06e3dce5877d493c2bcb4016d2b0dd62e5)     -      stuff
 - [d073aba](../../commit/d073abab5fc34438a676411be587d6efef6809ce)     -      setup

src/tests/logger.js
const chalk = require("chalk");
const logged = [{name: "Initialize", value: false}, 
                {name: "Port", value: false}, 
                {name: "Server", value: false}, 
                {name: "Routes", value: false}, 
                {name: "Database", value: false}
            ];
        
const config = require("../../package.json");
if(config.update.toLowerCase() === "yes") {
    logged.push({name: "Update", value: false});
}
process.stdout.moveCursor(0, -1);
process.stdout.clearLine();
var errors = [];
var test = false; 
var okOn = "Online";
var errOff = "Starting";
var date;
if(process.argv.find((x) => x === "test")) {
    test = true;
    okOn = "OK";
}

function log(change, err) {
    if(!date) {
        date = new Date();
    }
    if(err) {
        errOff = "Error";
    }
    console.clear();
    if(logged.find((x) => x.name.toLowerCase() === change) && !err) {
        logged.find((x) => x.name.toLowerCase() === change).value = true;
    }
    if(config.log === "tech") {
    //logged.forEach(l => process.stdout.clearLine());
    logged.forEach((l) => {
        process.stdout.moveCursor(0, -1);
        process.stdout.clearLine();
    });
    logged.forEach((l) => {
        if(l.value === true) {
            process.stdout.clearLine(); process.stdout.write(chalk.black.bgGreen.bold(" " + okOn + " ") + " " + l.name + "\n");
        } else {
            process.stdout.clearLine(); process.stdout.write(chalk.black.bgRed.bold(" " + errOff + " ") + " " + l.name + "\n");
        }
    });
    if(err) {
        if(!err.adv) {
            err.adv = "";
        }
        if(test !== true) {
            console.error("\nError: "  + err + " " + err.adv);
            return process.exit(0);
        } else {
            logged.find((x) => x.name.toLowerCase() === change.toLowerCase()).value = "err";
            errors.push(err);
        }
    }
    if(logged.filter((x) => x.value === false).length === 0) {
        if(errors.length > 0) {
            errors.forEach((e) => console.error("\nError: "  + e + " " + e.adv));
            return process.exit(1);
        }
        if(test) {
            process.stdout.write(chalk.bold.green("\nAll checks finished in " + (Math.round(((new Date() - date) / 1000) * 100) /100) + " seconds.\n\n"));
            return process.exit(0);
        }
        process.stdout.write(`\n${config.name}-${config.version} listening on port ${config.port}\n`);
        }
    } else {
        if(logged.filter((x) => x.value === false).length === 0) {
            process.stdout.write(`\n${config.name}-${config.version} listening on port ${config.port}\n`);   
        }
    }
}

if(test) {
    logged.push({name: "Syntax", value: false});
    require("child_process").exec("find .  -path ./node_modules -prune -o -path ./.history -prune -o -path ./data -prune -o -name \"*.js\" -exec node -c {} \\;", function(err, out) {
        if(err) {
            log("syntax", err);
        } else {
            log("syntax");
        }
    });
}

log("start");


module.exports = log;

src/public/files/test.js
document;

document.body;

/*
This should all minify
*/


src/utils/handlers/custom.js
const fs = require("file-system");
const Jimp = require("jimp");
const path = require("path");

Object.prototype.reply = function(code, message) {
  if (!this.status) throw new Error("This function is only for express!");
  this.status(code)
    .json({ status: code, err: message })
    .end();
};

Object.prototype.file = function(name, query) {
  if (!res.status) throw new Error("This function is only for express!"); // stop other object use

  const res = this; //Grab the result

  const location = path.resolve(__dirname + "/../../public/files/" + name); //Get the path

  fs.exists(location, function(exists) {
    if (!exists) return res.reply(404, "File not found!"); //Check for file

    const format = name.split(".")[1]; // Get the file type

    if (!/(gif|jpg|jpeg|tiff|png)$/i.test(format))
      return res.sendFile(location); // if its not an image just send it

    Jimp.read(location, (err, process) => {
      // get the width in query or use original file dimensions
      const width = parseInt(query.w) || process.bitmap.width;
      const height = parseInt(query.h) || process.bitmap.height;
      const quality = parseInt(query.q) || 100;
      try {
        process.quality(quality).resize(width, height);
        process.getBuffer(Jimp.MIME_JPEG, function(err, buffer) {
          res.set("Content-Type", Jimp.MIME_JPEG);
          res.send(buffer);
        });
      } catch (err) {
        res.reply(500, err.message);
      }
    });
  });
};


src/utils/handlers/database.js
const mongoose = require("mongoose");
const config = require("../../../package");

const log = require("../../tests/logger");
mongoose.connect("mongodb://localhost:27017/" + config.name, {useNewUrlParser: true})
    .catch((err) => {
        err.adv = "\n\nTry running npm run mongo\n"; 
        log("database", err);
    })
    .then((O_o) => {
        if(mongoose.connection.readyState === 1) {
            log("database");
        }
    });

var Db = require("../models/document");

const functions = {
    getAll(cb) {
        Db
        .find({})
        .exec((err, docs) => {
            if (err) {
                return cb(err, false);
            }
            if(docs) {
                return cb(null, docs);
            } else {
                return cb(null, false);
            }
        });
    },


    set(key, value, cb) {
        if(!key || !value) {
            return false;
        }
        var newDoc = new Db({key, value});
        console.log(JSON.stringify(newDoc));
        newDoc.save();
        cb(true);
    },


    get(key, cb) {
        Db.findOne({key}, function(err, res)  {
            if (err) {
                return cb(err, false);
            }
            if(res) {
                return cb(null, res);
            } else {
                return cb(null, false);
            }
        });
    }
};

module.exports = functions;


src/server.js
const createError = require("http-errors"),
  express = require("express"),
  path = require("path"),
  session = require("express-session"),
  bodyParser = require("body-parser"),
  cookieParser = require("cookie-parser"),
  logger = require("morgan"),
  helmet = require("helmet"),
  log = require("./tests/logger"),
  compression = require("compression"),
  minify = require("express-minify");

require("./utils/handlers/database");
require("./utils/handlers/custom");

const indexRouter = require("./routes/index");
const cdnRouter = require("./routes/cdn");

const app = express();
log("server");
// view engine setup
app.set("views", path.join(__dirname, "views"));
app.set("view engine", "ejs");

const cooky = {
  secret: "work hard",
  resave: true,
  expires: new Date() * 60 * 60 * 24 * 7,
  saveUninitialized: true
};

app.set("trust proxy", 1);
app.use(helmet());
app.use(session(cooky));
app.use(logger("tiny"));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(compression());
app.use(
  minify({
    cache: true
  })
);
app.use(express.static(path.join(__dirname, "public")));

app.use("/", indexRouter);
app.use("/f", cdnRouter);
// catch 404 and forward to error handler
app.use(function(req, res, next) {
  next(createError(404));
});

// error handler
app.use(function(err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get("env") === "development" ? err : {};

  // render the error page
  res.reply(err.status || 500, err.message);
});

log("routes");

module.exports = app;
